### ISSUE TITLE
[rita] Payment loop

### ISSUE THREAD

**Issue by jtremback** [althea-mesh/althea_rs#26]
Early on,  it's:

```
Sent payment, Payment: PaymentTx { to: Identity { ip_address: V6(2001::3), eth_address: EthAddress 0xb794f5ea0ba39494ce839613fffba74279579268, mac_address: MacAddress("12:34:56:78:90:ab") }, from: Identity { ip_address: V6(2001::2), eth_address: EthAddress 0xb794f5ea0ba39494ce839613fffba74279579268, mac_address: MacAddress("12:34:56:78:90:ab") }, amount: Uint256 108600 }
```

By the end, it's:

```
Sent payment, Payment: PaymentTx { to: Identity { ip_address: V6(2001::3), eth_address: EthAddress 0xb794f5ea0ba39494ce839613fffba74279579268, mac_address: MacAddress("12:34:56:78:90:ab") }, from: Identity { ip_address: V6(2001::2), eth_address: EthAddress 0xb794f5ea0ba39494ce839613fffba74279579268, mac_address: MacAddress("12:34:56:78:90:ab") }, amount: Uint256 210064840098451479073795730667720 }
```

**Comment by jtremback** [comment]
There's also some kind of feedback loop: https://gist.github.com/jtremback/c54fab150038c7defa533622d3abba94

**Comment by kingoflolz** [comment]
I think the problem is that because the metering is not working properly (packets are not counted by the sender), the payments are sometimes sent with no debt on the other side, in which case they are just sent right back.

**Comment by jtremback** [comment]
Oh, I forgot about the metering. I think that there may also be a conceptual issue with payments. 

Right now, nodes have one `debt` number for each of their neighbors. If the `debt` is above a threshold, they make a payment. The intent here is that the debt be added to as packets are forwarded for the neighbor.

However, the debt is also added to when receiving a payment. This is mostly intended to move the neighbor's debt upwards towards zero if it is in the negative, eventually resulting in a reopened tunnel if the neighbor pays off their debt..

What may end up happening is that

**Comment by jtremback** [comment]
Ah, I think the issue is that the `debt` is not decreased as the payment is made.

**Comment by jtremback** [comment]
The above fixed the increase, but not the loop.

**Comment by jtremback** [comment]
Talking with @ttk2, putting payments on some kind of cycle might be good. An option:

Store a map of `incoming_payments: HashMap<Identity, Uint256>`, and a map of `debts: HashMap<Identity, Int256>`.

When a payment comes in, add it to `incoming_payments`. Multiple payments coming in just increase this number, and do not trigger new payments.

When the traffic watcher comes back with the traffic totals for each neighbor, the traffic totals and the incoming payments are added into the debts, and then action is taken on the debts.

This will at least slow the loop down to once per traffic watcher cycle, which is not a huge issue.

**Comment by kingoflolz** [comment]
Another way of doing it would be to have make_payment in payment_controller keep a hashmap of "pending" payments, and execute them when it receives a "execute payment" packet or something.

Think they are conceptually pretty similar though.

**Comment by jkilpatr** [comment]
So the goal here is a loop, the problem is keeping the loop under control. I'm a big fan of PID control methods, but while they are cheap to build analog they are expensive to run in software. 

What's needed is a cycle system to keep the loop from running as quickly as we can get scheduled times on the processor and a sync system so that we can zero out balances on both sides and prevent unending growth of balances due to packet loss and other hardware realities.

**Comment by kingoflolz** [comment]
closed by #39
